elastic/logstash
# Reimplement LogStash::String setting in Java

## Release notes

[rn:skip]

## What does this PR do?

Reimplements `LogStash::Setting::String` Ruby setting class into the `org.logstash.settings.StringSetting` and exposes it through `java_import` as `LogStash::Setting::StringSetting`.
Updates the rspec tests in two ways:
- logging mock is now converted to real Log4J appender that spy log line that are later verified
- verifies `java.lang.IllegalArgumentException` instead of `ArgumentError` is thrown because the kind of exception thrown by Java code, during verification.

## Why is it important/What is the impact to the user?

As a developer I want use Java classes to shape the configuration settings.

## Checklist




## Author's Checklist


- [x] Implements the other constructors, with 2 and 3 params and spread the use
  - [x] Execute `SPEC_OPTS="-fd -P logstash-core/spec/logstash/runner_spec.rb" ./gradlew --rerun-tasks :logstash-core:rubyTests` there is a problem in reading the deprecation logger messages in https://github.com/elastic/logstash/blob/467ab3f44b6f0c2b09f6c39b7a5f90727e761485/logstash-core/spec/logstash/runner_spec.rb#L269. Suggestion mock the `Appender` and reconfigure logs?
- [x] Check the code for other instantiations of `LogStash::String.new`
- [x] Translate RSpec tests to JUnit if there are or cover with some testing
- ~~[ ] Rename `StringSetting` to `String` and use explicit `java.lang.String` around~~ converting means to make explicit reference to `java.lang.String` in all places inside the package `org.logstash.settings.*` .

## How to test this PR locally

- run of CI is one good test.
- run Logstash instance, it uses plenty of string settings.
- switch some string setting to an invalid value and verify it reports the error. For example tinker with `config.field_reference.escape_style` in `config/logstash.yml`

## Related issues


- Relates #15679
- Closes #16672
